File: source/downloads/code/LeetCode/Python/H-Index.py
class Solution(object):
    def hIndex(self, citations):
        """
        :type citations: List[int]
        :rtype: int
        """
        if citations == None or len(citations) == 0:
            return 0

        citations = sorted(citations)
        start = 0
        end = len(citations)-1

        while start <= end:
            mid = (start+end)/2
            if citations[mid] < len(citations)-mid:
                start = mid+1
            else:
                end = mid-1
        return len(citations)-start

File: source/downloads/code/LeetCode/Java/H-Index.java
public class Solution {
    public int hIndex(int[] citations) {
        int n = citations.length;
        // array2: 一篇文章的引用 -> 出现次数
        int[] array2 = new int[n+1];
        if(n==0) return 0;

        for(int c: citations){
            if(c >= n) array2[n] ++;
            else
                array2[c]++;
        }
        int ret = 0;
        // 便利array2找答案
        for(int i=n; i>=0; i--){
            // ret: citation 为i的出现次数
            ret += array2[i];
            if(ret >= i) return i;
        }
        return 0;
    }
}
